File: c++/leetcode/563. Binary Tree Tilt.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  int tilt(TreeNode *, int &);
  int findTilt(TreeNode *root)
  {
    int sum = 0;
    tilt(root, sum);
    return sum;
  }
};

int Solution::tilt(TreeNode *root, int &sum)
{
  if (!root)
    return 0;

  auto left = tilt(root->left, sum);
  auto right = tilt(root->right, sum);
  sum += abs(left - right);

  return left + right + root->val;
}

File: javascript/leetcode/563. Binary Tree Tilt.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findTilt = function (root) {
  let sum = 0;

  const tilt = (root) => {
    if (!root) return 0;

    let left = tilt(root.left);
    let right = tilt(root.right);

    sum += Math.abs(left - right);

    return left + right + root.val;
  };

  tilt(root);

  return sum;
};
